backend/src/Edna/ExperimentReader/Parser.hs

```diff
@@ -0,0 +1,123 @@
+  ) where
+
+import Universum
+
+import qualified Data.ByteString.Lazy as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as M
+
+import Codec.Xlsx (CellValue(..), Worksheet(..), cellValue, ixCell, toXlsxEither, wsCells, xlSheets)
+import Data.List.Split (wordsBy)
+import Data.Text (replace)
+import Fmt (pretty)
+import Lens.Micro (ix, _Just)
+import Text.Read (readParen)
+
+import Edna.ExperimentReader.Error (ExperimentParsingError(..))
+import Edna.ExperimentReader.Types
+  (CellType(..), Parameter(..), ParameterType(..), PointYX(..), Signal(..), PlateUnit(..))
+import Edna.Web.Types (ExperimentalMeasurement(..))
+
+type ParserType a = Either ExperimentParsingError a
+
+-- | Based on specific CellType return its value or get and exception that cell is empty or its
+-- expected type does not satisfy actual type
+specificCellAt
+  :: Typeable a
+  => Worksheet
+  -> PointYX
+  -> CellType a
+  -> Either ExperimentParsingError a
+specificCellAt workSheet p@(PointYX position) cellType =
+  let cell = workSheet ^? ixCell position . cellValue . _Just in
+  case (cellType, cell) of
+    (_, Nothing) -> Left EmptyCell
+    (_, Just (CellText "")) -> Left EmptyCell
+
+    (CText, Just (CellText x)) -> Right x
+    (CDouble, Just (CellDouble x)) -> Right x
+
+    (CSignal, Just (CellDouble x)) -> Right $ Signal x False
+    (CSignal, Just (CellText
+      -- If signal value is in brackets then it is outlier
+      (readParen True reads . toString . replace "," "." -> ((d, "") : _))
+      )) -> Right $ Signal d True
+
+    (_, cv) -> Left $ UnexpectedCellType p cellType cv
+
+-- | Check that cell representation equals to some value
+cellSatisfy :: Worksheet -> Text -> PointYX -> Bool
+cellSatisfy workSheet value (PointYX position) = case workSheet ^? ixCell position . cellValue of
+  Just (Just x) -> pretty x == value
+  _ -> value == ""
+
+-- | Common function for splitting
+paramsSplit :: Worksheet -> ParameterType -> Int -> NonEmpty Int -> ParserType [Parameter]
+paramsSplit workSheet pType normalAxisPoint indexes = constructParams (last indexes) $
+    flip NE.filter indexes $ \c -> not $ cellSatisfy workSheet "" $ direction (normalAxisPoint, c)
+  where
+    direction :: (Int, Int) -> PointYX
+    direction points = case pType of
+      Target -> PointYX points
+      Compound -> PointYX $ swap points
+
+    -- | Construct params from their indexes
+    constructParams :: Int -> [Int] -> ParserType [Parameter]
+    constructParams _ [] = pure []
+    constructParams lastId [idx] = do
+      name <- specificCellAt workSheet (direction (normalAxisPoint, idx)) CText
+      pure [Parameter name (idx, lastId)]
+    constructParams lastId (id1 : id2 : ids) = do
+      name <- specificCellAt workSheet (direction (normalAxisPoint, id1)) CText
+      restParams <- constructParams lastId (id2 : ids)
+      pure $ Parameter name (id1, id2 - 1) : restParams
+
+parseExperimentXls :: L.ByteString -> ParserType [ExperimentalMeasurement]
```

Comment:
It's an important top-level function, so please add a comment to it, even if its semantics is quite intuitive.